src/hotspot/share/logging/logAsyncFlusher.cpp

```diff
@@ -0,0 +1,233 @@
+                    _buffer_max_size, AsyncLogBufferSize);
+}
+
+bool AsyncLogMapIterator::do_entry(LogFileOutput* output, uintx* counter) {
+  using dummy = LogTagSetMapping<LogTag::__NO_TAG>;
+  LogDecorations decorations(LogLevel::Warning, dummy::tagset(), output->decorators());
+  const int sz = 128;
+  char out_of_band[sz];
+
+  if (*counter > 0) {
+    jio_snprintf(out_of_band, sz, UINTX_FORMAT_W(6) " messages dropped...", *counter);
+    output->write_blocking(decorations, out_of_band);
+    *counter = 0;
+  }
+
+  return true;
+}
+
+void LogAsyncFlusher::writeback(const LinkedList<AsyncLogMessage>& logs) {
+  LinkedListIterator<AsyncLogMessage> it(logs.head());
+  while (!it.is_empty()) {
+    AsyncLogMessage* e = it.next();
+    e->writeback();
+    e->destroy();
+  }
+}
+
+void LogAsyncFlusher::flush(bool with_lock) {
+  LinkedListImpl<AsyncLogMessage, ResourceObj::C_HEAP, mtLogging> logs;
+
+  if (with_lock) { // critical area
+    // Caveat: current thread must not hold _tty_lock or other lower rank lockers.
+    // Cannot install ttyUnlocker here because flush() may be invoked before defaultStream
+    // initialization.
+    MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+    _buffer.pop_all(&logs);
+    AsyncLogMapIterator iter;
+    _stats.iterate(&iter);
+  } else {
+    // C++ lambda can simplify the code snippet.
+    _buffer.pop_all(&logs);
+    AsyncLogMapIterator iter;
+    _stats.iterate(&iter);
+  }
+
+  writeback(logs);
+}
+
+void LogAsyncFlusher::run() {
+  while (_state == ThreadState::Running) {
+    {
+      MonitorLocker m(&_lock, Mutex::_no_safepoint_check_flag);
+      m.wait(500 /* ms, timeout*/);
+    }
+    flush();
+  }
+
+  // Signal thread has terminated
+  MonitorLocker ml(Terminator_lock);
+  _state = ThreadState::Terminated;
+  ml.notify_all();
+}
+
+LogAsyncFlusher* LogAsyncFlusher::_instance = nullptr;
+
+void LogAsyncFlusher::initialize() {
+  if (!LogConfiguration::is_async_mode()) return;
+
+  if (_instance == NULL) {
+    Atomic::release_store(&LogAsyncFlusher::_instance, new LogAsyncFlusher());
+  }
+}
+
+// Termination
+// 1. issue an atomic release_store to close the logging window.
+// 2. flush itself in-place.
+// 3. signal asynclog thread to exit.
+// 4. wait until asynclog thread exits.
+// 5. (optional) delete this in post_run().
+void LogAsyncFlusher::terminate() {
+  if (_instance != NULL) {
+    LogAsyncFlusher* self = _instance;
+
+    Atomic::release_store<LogAsyncFlusher*, LogAsyncFlusher*>(&_instance, nullptr);
+    self->flush();
+    {
+      MonitorLocker ml(&self->_lock, Mutex::_no_safepoint_check_flag);
+      self->_state = ThreadState::Terminating;
+      ml.notify();
+    }
+    {
+      MonitorLocker ml(Terminator_lock, Mutex::_no_safepoint_check_flag);
+      while (self->_state != ThreadState::Terminated) {
+        ml.wait();
+      }
+    }
+  }
+}
+
+LogAsyncFlusher* LogAsyncFlusher::instance() {
+  if (Thread::current_or_null() != nullptr) {
```

Comment:
`LogAsyncFlusher::enqueue` needs Thread::current() to use _lock. 
```cpp
void Mutex::lock_without_safepoint_check() {
  lock_without_safepoint_check(Thread::current());
}
```

This check is for a very rare corner case.  This can be trigger using the following cmdline.
`
java -Xlog:'thread+smr=debug:file=/tmp/t.log' -Xlog:async -version
`
the root cause is the following code in thread

```cpp
void ThreadsSMRSupport::smr_delete(JavaThread *thread) {
  elapsedTimer timer;
  if (EnableThreadSMRStatistics) {
    timer.start();
  }

  wait_until_not_protected(thread);

  delete thread; // Thread::current() will trigger assert(current != __null) failed: Thread::current() called on detached thread after ~Thread()
  if (EnableThreadSMRStatistics) {
    timer.stop();
    uint millis = (uint)timer.milliseconds();
    ThreadsSMRSupport::inc_deleted_thread_cnt();
    ThreadsSMRSupport::add_deleted_thread_times(millis);
    ThreadsSMRSupport::update_deleted_thread_time_max(millis);
  }
  // logsite will crash  in async mode because we can't use Thread::current() here. 
  log_debug(thread, smr)("tid=" UINTX_FORMAT ": ThreadsSMRSupport::smr_delete: thread=" INTPTR_FORMAT " is deleted.", os::current_thread_id(), p2i(thread));
}
```

Reply:
Sigh. I get it. Another example that a good logging system should be independent from the infrastructure of the surrounding system. Any chance this patch could use raw locking primitives, pthread_mutex resp. CreateMutex, without having to rewrite everything? 

Ironically I though async logging would reduce the number of log call sites where the conditions for logging are not met, by moving most of the stack out of the caller into a separate thread. But the Thread::current condition is a new limitation then compared to sync logging.

If you don't want to change the Mutex:

So you have some rare log sites which won't allow you to use asynchronous logging since you cannot use locks, and you solve this by returning NULL, which then falls back to synchronous logging? What happens if you mix synchronous and asynchronous logging like this? Does this work?

From a code point of view your reasoning is very difficult to deduce. Can you please change this: remove the THread::current condition from the singleton getter, and add an explicit check for Thread::current at its caller site. In async mode either just omit printing any output needing Thread::current, or relegate to synchronous printing. And add a good comment there.

Reply:
Thank you for reviewing. you raise the exact problem which keep me from sleeping...

> From a code point of view your reasoning is very difficult to deduce. Can you please change this: remove the THread::current condition from the singleton getter, and add an explicit check for Thread::current at its caller site. In async mode either just omit printing any output needing Thread::current, or relegate to synchronous printing. And add a good comment there.

That's exactly the current approach. It works but async logging and synchronous logging may overlap in some rare windows such as my example of `smr_delete`. 

```cpp
LogAsyncFlusher* LogAsyncFlusher::instance() {
  // thread may has been detached, then Thread::current() used by Mutex is inavailable.
  // eg. ThreadsSMRSupport::smr_delete() uses log_debug() after delete thread.
  if (Thread::current_or_null() != nullptr) {
    return _instance;
  } else {
    return nullptr;
  }
}
```

One drawback is that I have to give up the  FileLocker optimization for the entire interval async logging(99.99% time)
for those "rare" windows :( Once I remove FileLocker for FILE*, it may get stuck when two logging writers overlap!

I am evaluating different approaches.  eg. the 2-synchronization-points design or even replace Mutex with a semaphore-based lock like [ConfigurationLock](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/logging/logConfiguration.cpp#L55)